### ISSUE TITLE
Feature Request: Correlation Layer 

### ISSUE THREAD

**Issue by atilaorh** [tensorflow/tensorflow#11956]
There is a surging interest in Geometric Computer Vision and a large number of recent papers leveraging an operation(with small variations) dubbed "Correlation" layer. There is a CUDA kernel for this operation in the FlowNet paper's author's fork of Caffe. Is there plan on including it in Tensorflow? I couldn't locate a relevant issue anywhere and this is why I am raising this issue.

**Comment by shivaniag** [comment]
@benoitsteiner do you have any comment on this?

**Comment by yaroslavvb** [comment]
Maybe this could be contributions welcome? First step could be implemented this as a user_op which could exist in a separate repo

**Comment by 9thDimension** [comment]
+1 for this request.

If anyone's interested there's a working (albeit slightly crude) implementation using Keras over Tensorflow here:

https://github.com/WeidiXie/New_Layers-Keras-Tensorflow/blob/master/src/normalized_correlation_layer.py

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by JiamingSuen** [comment]
Just FYI: There are also two implementations of correlation layer on Github:
1. https://github.com/sampepose/flownet2-tf/tree/master/src/ops/correlation
2. https://github.com/jgorgenucsd/corr_tf

**Comment by mtngld** [comment]
+1,

Is it possible to implement this using current `conv3d` layer? The doc for it says:
"Our Conv3D implements a form of cross-correlation.", but I was not able to make it actually work for correlating 2 images.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by drpngx** [comment]
Sounds like a good potential contribution!

**Comment by Arsalan-zafar** [comment]
+1

**Comment by djl11** [comment]
ditto, would be very useful.

**Comment by zboldyga** [comment]
Thanks for the helpful links to user-level solutions! I'll start implementing this later in July, and given my free time it will likely take me till around September to complete.

**Comment by evolu8** [comment]
@zboldyga , did you get anywhere? This would be very useful. Something a little better engineered, documented and generic than some of the efforts already out there would be wonderful.

**Comment by zboldyga** [comment]
@evolu8 eek! I totally lost track of this ticket, thanks for the reminder... I read the paper today as a refresher and the correlation layer seems straightforward. Will be in the weeds and make some commits shortly 😄

**Comment by evolu8** [comment]
Brilliant. Good luck!

**Comment by seanpmorgan** [comment]
This can be closed as it was added in https://github.com/tensorflow/addons/pull/207

cc @facaiy

**Comment by facaiy** [comment]
Well done
